dll/win32/shlwapi/autocomp.cpp

```diff
@@ -0,0 +1,623 @@
+        return TRUE;
+    }
+    return FALSE;
+}
+
+static DWORD GetDriveFlags(LPCWSTR pszRoot)
+{
+    WCHAR szDevice[16];
+    HANDLE hDevice;
+    ULONG ret = 0;
+
+    StringCbCopyW(szDevice, sizeof(szDevice), L"\\\\.\\");
+    szDevice[4] = pszRoot[0];
+    szDevice[5] = L':';
+    szDevice[6] = 0;
+    hDevice = CreateFileW(szDevice, FILE_READ_ATTRIBUTES,
+                          FILE_SHARE_READ | FILE_SHARE_WRITE,
+                          NULL, OPEN_EXISTING, 0, NULL);
+    if (hDevice == INVALID_HANDLE_VALUE)
+        return ret;
+
+    GetDriveCharacteristics(hDevice, &ret);
+    CloseHandle(hDevice);
+    return ret;
+}
+
+static BOOL IsSlowDrive(LPCWSTR pszRoot)
+{
+    DWORD ret;
+    switch (GetDriveTypeW(pszRoot))
+    {
+        case DRIVE_REMOVABLE:
+            ret = GetDriveFlags(pszRoot);
+            if (!(ret & FILE_FLOPPY_DISKETTE) || (ret & FILE_VIRTUAL_VOLUME))
+                break;
+            return TRUE; // Floppy and non-virtual
+
+        case DRIVE_CDROM:
+            ret = GetDriveFlags(pszRoot);
+            if (ret & FILE_VIRTUAL_VOLUME)
+                break;
+            return TRUE; // CD/DVD and non-virtual
+    }
+    return FALSE; // Not slow
+}
+
+BOOL CAutoCompleteEnumString::DoFileSystem(LPCWSTR pszQuery)
+{
+    // Check the drive
+    INT nDriveNumber = PathGetDriveNumberW(pszQuery);
+    if (nDriveNumber != -1)
+    {
+        WCHAR szRoot[] = L"C:\\";
+        szRoot[0] = WCHAR('A' + nDriveNumber);
+        assert(PathIsRootW(szRoot));
+        if (IsSlowDrive(szRoot))
+            return FALSE; // Don't scan slow drives
+    }
+
+    // Is it directory-only?
+    BOOL bDirOnly = !!(m_dwSHACF & SHACF_FILESYS_DIRS);
+
+    DWORD attrs = GetFileAttributesW(pszQuery);
+    if (attrs != INVALID_FILE_ATTRIBUTES) // File or folder does exist
+    {
+        if (attrs & FILE_ATTRIBUTE_DIRECTORY) // Is it a directory?
+        {
+            size_t cch = wcslen(pszQuery);
+            if (cch > 0 && pszQuery[cch - 1] == L'\\') // The last character is '\\'
+            {
+                DoDir(pszQuery, bDirOnly); // Scan the directory
+                return FALSE; // Not exact match
+            }
+        }
+        return TRUE; // Exact match should hide the list
+    }
+    else if (pszQuery[0] && wcschr(pszQuery, L'\\') != NULL)
+    {
+        // Non-existent but can be a partial path
+        CStringW strPath = pszQuery;
+
+        // Remove the file title part
+        INT ich = strPath.ReverseFind(L'\\');
+        if (ich >= 0)
+            strPath = strPath.Left(ich);
+
+        DoDir(strPath, bDirOnly); // Scan the directory
+    }
+    else
+    {
+        DoDrives(bDirOnly); // Scan drives
+    }
+
+    return FALSE; // Not exact match
+}
+
+void CAutoCompleteEnumString::DoTypedPaths(LPCWSTR pszQuery)
+{
+    static const LPCWSTR
+        pszTypedPaths = L"Software\\Microsoft\\Windows\\CurrentVersion\\Explorer\\TypedPaths";
+    WCHAR szName[32], szValue[MAX_PATH + 32];
```

Comment:
We should remove `MAX_PATH` of here?

Reply:
If it is not making this very complex, yes please.

Reply:
OK. Done.